### ISSUE TITLE
 save only the best model and weights after training

### ISSUE THREAD

**Issue by shivahanifi** [shivahanifi/SCDD-image-segmentation-keras#17]
The `train_script.py`, using `CheckpointsCallback` saves the weights for each of the epochs in `.data` and `.index` format. In issue #15 I made it also save the model in `.h5` format, but it is saving all the checkpoints in that format and with an inefficient naming convention. 
## To Do: save only the best model and weights after training  with a `{wandb.run.name}_model.h5` format.

**Comment by shivahanifi** [comment]
## Solved
Using the following code I managed to save the best model with the preferred naming convention.
```
from keras.callbacks import Callback, ModelCheckpoint

# Create ModelCheckpoint callback to save only the best model based on validation metric
checkpoint_callback = ModelCheckpoint(
    checkpoint_path,
    monitor="accuracy", 
    save_best_only=True,
    save_weights_only=False,  # Save entire model, not just weights
    verbose=1
)
```
```
# After training, save the best model as an .h5 file
best_model_h5_path = os.path.join(checkpoint_path, f"{wandb.run.name}_best_model.h5")
model.save(best_model_h5_path)
artifact = wandb.Artifact(name=wandb.run.name, type="model")
artifact.add_dir(checkpoint_path)
wandb.log_artifact(artifact)
```

you will have the TensorFlow's SavedModel format in the checkpoint path containing the following files:
|file|purpose|
|---|---|
|assets directory| to store any assets needed for your model that are not included in the computation graph itself (vocabularies, label maps, or other auxiliary files required during model inference or evaluation.)|
|variables directory| stores the model's variables (weights and biases) in a structured format. The `.data` file contains the actual variable values (weights), while the `.index` file contains metadata about those variables, allowing TensorFlow to correctly load them.|
| `_config.json`| configuration information for the model |
| `fingerprint.pb`|  to provide a unique fingerprint for the model.|
| `keras_metadata.pb`| metadata related to the Keras model. includes information about the model architecture, training configuration (like loss functions and metrics), and some additional data about the model's versioning and structure. This file is particularly useful for loading the model in Keras and maintaining compatibility across different versions of Keras.|
| `saved_model.pb`| main file that describes the model's computation graph. contains the complete TensorFlow computation graph, including all operations and their relationships. This file is essential for restoring the model to its original state during inference or further training.|
| `{wandb.run.name}_model.h5`| an HDF5 file that stores a complete Keras model(a self-contained representation of the model) . It contains the model architecture, weights, optimizer state, and training configuration, all in a single file. This format is useful for quick loading and is widely supported in Keras.|

**Comment by shivahanifi** [comment]
For network visualizations with Netron both the `{wandb.run.name}_model.h5`  and `keras_metadata.pb` can be used. However,  `{wandb.run.name}_model.h5`  contains detailed information regarding the architecture.
|`keras_metadata.pb` |`{wandb.run.name}_model.h5` |
|---|---|
|![Screenshot from 2024-10-30 14-22-18](https://github.com/user-attachments/assets/a9fdc329-7dbd-49a7-a903-c06852efaa07)|![Screenshot from 2024-10-30 14-20-32](https://github.com/user-attachments/assets/b5ee4d49-4704-4cba-aef9-bd8d29c4ba24)|

![Screenshot from 2024-10-30 14-59-29](https://github.com/user-attachments/assets/b09d9e28-7eac-41c5-a606-46ebdcf166cb)

**Comment by shivahanifi** [comment]
## saving multiple epochs

```
# Create ModelCheckpoint callback to save only the best model based on validation metric
checkpoint_callback = ModelCheckpoint(
    os.path.join(checkpoint_path, "model_epoch_{epoch:03d}"),
    save_best_only=False,
    save_weights_only=False,  # Save entire model, not just weights
    save_freq="epoch", # Save model at the end of the epoch if it is the best
    verbose=1,
)

```

**Comment by shivahanifi** [comment]
Saving the best model is solved in this issue. issue #18 will focus on logging the info related to the best model.
